src/Hosting/Hosting/src/Internal/HostingMetrics.cs

```diff
@@ -0,0 +1,174 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Metrics;
+
+namespace Microsoft.AspNetCore.Hosting;
+
+internal sealed class HostingMetrics : IDisposable
+{
+    public const string MeterName = "Microsoft.AspNetCore.Hosting";
+
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentRequestsCounter;
+    private readonly Histogram<double> _requestDuration;
+
+    public HostingMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.CreateMeter(MeterName);
+
+        _currentRequestsCounter = _meter.CreateUpDownCounter<long>(
+            "current-requests",
+            description: "Current requests");
+
+        _requestDuration = _meter.CreateHistogram<double>(
+            "request-duration",
+            description: "Request duration");
+    }
+
+    public void RequestStart(string scheme, string method, HostString host)
+    {
+        // Tags must match request end.
+        var tags = CreateRequestTags(scheme, method, host);
+        _currentRequestsCounter.Add(1, tags);
+    }
+
+    public void RequestEnd(string scheme, string method, HostString host, string? routePattern, int statusCode, Exception? exception, IList<KeyValuePair<string, object?>>? customTags, long startTimestamp, long currentTimestamp)
+    {
+        // Tags must match request start.
+        var tags = CreateRequestTags(scheme, method, host);
+        _currentRequestsCounter.Add(-1, tags);
+
+        // Add information gathered during request.
```

Comment:
Shouldn't we avoid all the processing beyond this point if the `_requestDuration` is disabled

Reply:
There are checks outside this method to ensure the method is only called if metrics/diagnostics are enabled.

Reply:
+1

Reply:
I see 2 checks
- IsEnabled which is checking for if either `_currentRequestsCounter` or `_requestDuration` is enabled then this processing gets enabled. So, even if someone has only enabled `currentRequestCounter`, they will have to pay the processing cost that follows here for emitting `_requestDuration`. This is just not right.
- The other checks would be whether there is a listener for the metric source which will be true because the source for both these metrics are same.

I just don't want dotnet to repeat the same mistakes it did with http request logging where it's just enabled by default and customers have no easy and non-hacky way of disabling this

Reply:
Added individual if checks.